### DISCUSSION TITLE
alembic check detects upgrade operations but flask-migrate doesn't generate new revision (v.2.5.3)

### DISCUSSION THREAD

**Discussion by HusseinWarsame** [miguelgrinberg/Flask-Migrate#507] [Q&A]
Hello,

My team has a few redundant operations in the revision history of some our projects. I'd like to remove them since it is conflicting with the adoption of a naming convention but have encountered some unexpected errors.

For example,

![image](https://user-images.githubusercontent.com/21069829/219738275-f239b05d-56be-43a6-a312-a24d018d3937.png)

Here we have the addition and removal of the same foreign key on two different tables. My expectation was that the "sum" of these operations would not count as a pending upgrade operation. This is because when I try to generate a migration with Flask-Migrate for these pending upgrade operations, there are no changes in the schema detected:

![image](https://user-images.githubusercontent.com/21069829/219740299-4edd1199-0d27-4944-b597-b6a9d74ba666.png)

This doesn't seem like expected behaviour [according to the alembic documentation](https://alembic.sqlalchemy.org/en/latest/autogenerate.html#running-alembic-check-to-test-for-new-upgrade-operations)

> ...[alembic check] returns an error code plus a message if it is detected that new operations would be rendered into a new revision, or if not, returns a success code plus a message. When alembic check returns a success code, this is an indication that the alembic revision --autogenerate command would produce only empty migrations, and does not need to be run.

I've raised a [similar discussion with the alembic maintainers](https://github.com/sqlalchemy/alembic/discussions/1179) but they've deemed it as expected behaviour. If there's anything on the flask-migrate side of things that I'm misunderstanding, please let me know.

Thanks

**Comment by miguelgrinberg** [comment]
Do you have the actual details included in the `AutogeneratedDiffsDetected`? In spite of what Mike told you, this does look like an Alembic bug. Those remove+add of the same key are probably renames, my guess is that Alembic does not like the name in those constraints and tries to make new ones. If that is the case in the check side, then I think it is definitely a bug that the autogenerate does not include those.

**Reply by HusseinWarsame** [reply]
Is this what you mean by the actual details? 

![image](https://user-images.githubusercontent.com/21069829/219802504-cf559f13-1017-486f-b580-c9b309bf6e81.png)

Yes, you are correct. In this case, the remove+add are renames caused by the addition of a naming convention.
However, I've also encountered this issue without imposing a naming convention. This is what happened in another one of our projects. 

![image](https://user-images.githubusercontent.com/21069829/219804545-ef621bbe-d41b-4a79-8624-0432d90b085b.png)
![image](https://user-images.githubusercontent.com/21069829/219805224-0f6b0663-0c23-4791-969f-d17db62d0119.png)

The only pending upgrade operation that `check` detects here is a removal of a table. But once again, no new revision is generated with `flask db migrate`

![image](https://user-images.githubusercontent.com/21069829/219804647-0dcd33e0-28a2-4d2d-9f63-e66d0afc88e1.png)

**Reply by miguelgrinberg** [reply]
It's impossible for me to tell you what's wrong here, but my guess is that this is user error (i.e. your error). From the screenshots it seems that you are running the `check` command as part of some sort of scripted action, but you are running the `migrate` command directly from the CLI, and on a Docker container (unclear if the scripted event was running on same Docker container or if it is a container at all).

What I think you need to do is create an example that you can provide to Mike that shows a "alembic check" followed by "alembic revision --autogenerate" showing inconsistent results. The difference in the way you run these commands is likely the reason you are getting different results. There is something in the database (or the code, I guess) that must be different between the two commands.
